sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala

```diff
@@ -0,0 +1,237 @@
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val inputTypeCheck = super.checkInputDataTypes()
+    if (inputTypeCheck.isSuccess) {
+      val formatStringValid = formatString.forall {
+        case 'x' | 'X' | '-' => true
+        case ch if ch.isDigit || ch.isWhitespace => true
+        case _ => false
+      }
+      if (formatStringValid) {
+        TypeCheckResult.TypeCheckSuccess
+      } else {
+        throw QueryCompilationErrors.maskCcnInvalidFormatError(prettyName)
+      }
+    } else {
+      inputTypeCheck
+    }
+  }
+
+  override def nullSafeEval(string: Any, format: Any): Any =
+    parser.parse(string.asInstanceOf[UTF8String])
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val builder = ctx.addReferenceObj(
+      "parser", parser, classOf[MaskDigitSequenceParser].getName)
+    val eval = left.genCode(ctx)
+    ev.copy(code =
+      code"""
+        |${eval.code}
+        |boolean ${ev.isNull} = ${eval.isNull};
+        |${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+        |if (!${ev.isNull}) {
+        |  ${ev.value} = $builder.parse(${eval.value});
+        |  if (${ev.value} == null) {
+        |    ${ev.isNull} = true;
+        |  }
+        |}
+      """.stripMargin)
+  }
+}
+
+/** Executes the string parsing steps for the MaskDigitSequence class. */
+class MaskDigitSequenceParser(formatString: String, nullOnError: Boolean) extends Serializable {
+  def parse(input: UTF8String): UTF8String = {
+    val inputString = input.toString
+    var formatStringIndex = 0
+    var error = false
+    def skipFormatWhitespace(): Unit =
+      while (formatStringIndex < formatString.length &&
+        formatString(formatStringIndex).isWhitespace) {
+        formatStringIndex += 1
+      }
+    // Check and consume each character in the input string, comparing against characters in the
+    // format string.
+    val result = inputString.map { inputChar =>
+      if (error) {
+        // If we have encountered an error, leave the input character alone; we will raise an
+        // exception or return NULL after this loop has finished.
+        inputChar
+      } else if (inputChar.isWhitespace) {
+        // The input character is whitespace. Ignore it and continue comparing the next input
+        // character against the same format string character as this iteration.
+        inputChar
+      } else if (formatStringIndex >= formatString.length) {
+        // We have already consumed all the characters in the format string, but one or more
+        // characters exist in the input string. This is an error because the input string does not
+        // match the format string.
+        error = true
+        'X'
+      } else {
+        // Check the corresponding character in the format string.
+        skipFormatWhitespace()
```

Comment:
Seems we are going to skip the spaces on both input/format string...

Reply:
Same, this is per intention (left a comment).

Reply:
Update on this: per our offline conversation, we decided to require whitespace in the format string to match the input string exactly. I updated the PR accordingly.